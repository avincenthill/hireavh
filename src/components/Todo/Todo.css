.todo {
  margin: calc(var(--s-xs) / 2);
  background-color: var(--c-black);
  width: var(--s-xl);
  height: var(--s-m);
  color: var(--c-1);
  display: flex;
  align-items: center;
  align-content: center;
  justify-content: space-between;
  border-radius: var(--s-xs);
  box-shadow: inset 0 0 0 var(--b-1) var(--c-1);
  box-sizing: border-box;
}

.todo-reset {
  justify-content: center;
}

.todo > * {
  pointer-events: none;
}

.todo-true {
  color: white;
  background-color: green;
}

.todo-container {
  margin-bottom: var(--s-l);
}

.todo-title {
  font-size: var(--fs-m);
  margin-bottom: var(--s-xs);
}

.todo-icon {
  margin-left: var(--s-s);
}

.todo-item {
  font-size: var(--fs-s);
  margin-left: var(--s-xs);
  color: var(--c-white);
  margin-right: var(--s-s);
  font-weight: bold;
}

.todo-item-reset {
  font-size: var(--fs-s);
  margin-left: var(--s-xs);
  color: var(--c-white);
  font-weight: bold;
}

.todo-button {
  background-color: var(--c-4);
  border-radius: 0 var(--s-xs) var(--s-xs) 0;
  border: var(--c-1) solid var(--b-1);
}

.todo-a {
  font-size: var(--fs-xs);
  text-decoration: none;
  color: var(--c-1);
}

.todo-resetBtn {
  margin-top: var(--s-m);
}

:root {
  --noise-dist: 1vmin;
  --noise-time: 1vmin;
}

.confetti {
  width: 5vmin;
  height: 5vmin;
  background-color: #f2d74e;
  position: absolute;
  left: 50%;
  animation: confetti 2s ease-in-out -1s infinite;
  transform-origin: left top;
  z-index: var(--z-top);
}
.confetti:nth-child(1) {
  background-color: #f2744e;
  left: 10%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(2) {
  background-color: #0f84c9;
  left: 20%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(3) {
  background-color: #ff9a91;
  left: 30%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(4) {
  background-color: #f2d74e;
  left: 40%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(5) {
  background-color: #95c3de;
  left: 50%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(6) {
  background-color: #ff9a91;
  left: 60%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(7) {
  background-color: #d60bb4;
  left: 70%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(8) {
  background-color: #95c3de;
  left: 80%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(9) {
  background-color: #ff9a91;
  left: 90%;
  animation-delay: var(--noise-time);
}
.confetti:nth-child(10) {
  background-color: #05d320;
  left: 100%;
  animation-delay: var(--noise-time);
}

@keyframes confetti {
  0% {
    transform: rotateZ(15deg) rotateY(0deg) translate(0, 0);
  }
  25% {
    transform: rotateZ(5deg) rotateY(360deg)
      translate(calc(-5vw + var(--noise-dist)), calc(20vh + var(--noise-dist)));
  }
  50% {
    transform: rotateZ(15deg) rotateY(720deg)
      translate(calc(5vw + var(--noise-dist)), calc(60vh + var(--noise-dist)));
  }
  75% {
    transform: rotateZ(5deg) rotateY(1080deg)
      translate(calc(-10vw + var(--noise-dist)), calc(80vh + var(--noise-dist)));
  }
  100% {
    transform: rotateZ(15deg) rotateY(1440deg)
      translate(calc(10vw + var(--noise-dist)), calc(100vh + var(--noise-dist)));
  }
}
