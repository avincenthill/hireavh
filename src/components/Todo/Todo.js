import './Todo.css';
import {
  FaBookOpen,
  FaBroom,
  FaCheckCircle,
  FaCode,
  FaGlassCheers,
  FaHandPaper,
  FaMailBulk,
  FaPills,
  FaRunning,
  FaTooth,
  FaTrophy,
  FaTshirt,
  FaUserFriends,
} from 'react-icons/fa';
import React, { Component } from 'react';
import { IconContext } from 'react-icons';
import Page from 'components/Page/Page';
import checkSound from 'assets/sounds/check.mp3';
// import content from 'content/content'; // TBD pull out page content
import resetSound from 'assets/sounds/reset.mp3';
import styleconfig from 'styles/styleconfig';
import successSound from 'assets/sounds/success.mp3';
import utils from 'utils/utils';

export default class Todo extends Component {
  constructor(props) {
    super(props);
    function datesAreOnSameDay(first, second) {
      return (
        first.getFullYear() === second.getFullYear() &&
        first.getMonth() === second.getMonth() &&
        first.getDate() === second.getDate()
      );
    }

    // TBD: add condition tasks based on day of week
    const todoTaskArray = [
      {
        title: 'EXERCISE',
        icon: <FaRunning />,
      },
      {
        title: 'LEETCODE',
        icon: <FaCode />,
      },
      {
        title: 'EMAIL',
        icon: <FaMailBulk />,
      },
      {
        title: 'STUDY',
        icon: <FaBookOpen />,
      },
      {
        title: 'CLEAN APT',
        icon: <FaBroom />,
      },
      {
        title: 'DO DISHES',
        icon: <FaGlassCheers />,
      },
      {
        title: 'REACH OUT',
        icon: <FaUserFriends />,
      },
      {
        title: 'GET REJECTED',
        icon: <FaHandPaper />,
      },
      {
        title: 'TAKE MULTIVITAMIN',
        icon: <FaPills />,
      },
      {
        title: 'PREP TOMORROW',
        icon: <FaTshirt />,
      },
    ];

    this.state = {
      todoTaskArray,
    };

    const lastSeen = JSON.parse(localStorage.getItem('lastSeen'));
    if (!lastSeen || !datesAreOnSameDay(new Date(lastSeen), new Date())) {
      // clear data and reset seen date
      localStorage.setItem('lastSeen', JSON.stringify(new Date()));
      this.resetTodoArr();
    }

    let todoArr = this.getTodoArr();
    if (!todoArr || todoArr.length !== todoTaskArray.length) {
      this.resetTodoArr();
    }
  }

  componentDidMount() {
    this._mounted = true;
  }

  componentWillUnmount() {
    this._mounted = false;
  }

  getTodoArr = () => {
    return JSON.parse(localStorage.getItem('todoArr'));
  };

  handleResetClick = () => {
    this.playResetSound();
    this.resetTodoArr();
  };

  resetTodoArr = () => {
    const todoArr = Array(this.state.todoTaskArray.length).fill(false);
    localStorage.setItem('todoArr', JSON.stringify(todoArr));
    if (this._mounted) this.forceUpdate();
  };

  handleClick = (e) => {
    const todoArr = this.getTodoArr();
    const id = parseInt(e.target.id, 10);
    const newBool = !todoArr[id];
    todoArr[id] = newBool;
    localStorage.setItem('todoArr', JSON.stringify(todoArr));
    if (this._mounted) this.forceUpdate();
    this.handleTodoListState(newBool, todoArr);
  };

  handleTodoListState = (selectedBool, todoArrBool) => {
    const areSomeFalse = todoArrBool.some((element) => {
      return !element;
    });

    if (areSomeFalse) {
      if (selectedBool) {
        this.playCheckSound();
      } else {
        this.playResetSound();
      }
    } else {
      this.playCheckSound();
      this.animateSuccess();
    }
  };

  playSound = (soundElClass, volume) => {
    const audioEl = document.getElementsByClassName(soundElClass)[0];
    audioEl.volume = volume;
    audioEl.currentTime = 0;
    audioEl.play();
  };

  playResetSound = () => {
    this.playSound('resetSound', 0.1);
  };

  playSuccessSound = () => {
    this.playSound('successSound', 0.6);
  };

  playCheckSound = () => {
    this.playSound('checkSound', 0.5);
  };

  animateSuccess = async () => {
    const todoArr = this.getTodoArr();
    const todoArrLen = todoArr.length;
    const delay = 50;

    for (let i = 0; i < todoArrLen; i += 1) {
      todoArr[i] = false;
      localStorage.setItem('todoArr', JSON.stringify(todoArr));
      if (this._mounted) this.forceUpdate();
      await utils.sleep(delay);
    }

    for (let i = 0; i < todoArrLen; i += 1) {
      todoArr[i] = true;
      localStorage.setItem('todoArr', JSON.stringify(todoArr));
      if (this._mounted) this.forceUpdate();
      await utils.sleep(delay);
    }

    this.playSuccessSound();
  };

  renderArrayOfTodos() {
    const todoArr = this.getTodoArr();

    return todoArr.map((todoBool, index) => {
      return (
        <button
          id={index}
          key={index}
          onClick={this.handleClick}
          className={`todo button-hover-light todo-${todoBool}`}
        >
          <span className="todo-icon">
            {!todoBool ? (
              this.state.todoTaskArray[index].icon
            ) : (
              <FaCheckCircle />
            )}
          </span>
          <span className="todo-item">
            {this.state.todoTaskArray[index].title}
          </span>
        </button>
      );
    });
  }

  render() {
    const todoArr = this.getTodoArr();
    const areSomeFalse = todoArr.some((element) => {
      return !element;
    });

    return (
      <div className="todolist-container">
        <Page>
          <IconContext.Provider
            value={{
              style: {
                ...styleconfig.icons.l,
                ...utils.getIconStyles('todo'),
              },
            }}
          >
            <h3 className="todo-title">{new Date().toDateString()}</h3>
            {this.props ? (
              <div className="todo-container">{this.renderArrayOfTodos()}</div>
            ) : null}
            {!areSomeFalse ? (
              <span className="todo-trophy">
                <IconContext.Provider
                  value={{
                    style: {
                      ...styleconfig.icons.xl,
                    },
                  }}
                >
                  <FaTrophy />
                </IconContext.Provider>
              </span>
            ) : null}
            <button
              id="2"
              className="todo todo-reset button-hover-light todo-reset-btn"
              onClick={this.handleResetClick}
            >
              <span className="todo-item-reset">RESET</span>
              <audio className="checkSound">
                <source src={checkSound}></source>
              </audio>
              <audio className="successSound">
                <source src={successSound}></source>
              </audio>
              <audio className="resetSound">
                <source src={resetSound}></source>
              </audio>
            </button>
            <h3 className="todo-title todo-bottom-date">
              {new Date().toDateString()}
            </h3>
          </IconContext.Provider>
        </Page>
      </div>
    );
  }
}
