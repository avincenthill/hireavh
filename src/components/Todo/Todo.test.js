import React from 'react';
import Todo from './Todo';
import renderer from 'react-test-renderer';

describe('Todo', () => {
  describe('snapshot', () => {
    it('should match the last saved snapshot', () => {
      const tree = renderer.create(<Todo />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
