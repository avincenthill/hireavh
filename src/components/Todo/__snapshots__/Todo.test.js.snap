// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo snapshot should match the last saved snapshot 1`] = `
<div
  className="todo-container"
>
  <div>
    <div>
      <div
        className="nav nav-top"
      >
        <div
          className="nav-icon $
          this.props.isBottom ? \\"nav-icon-bottom\\" : \\"nav-icon-top\\"
        }"
        >
          <svg
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-1)",
                "height": "3rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 8 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 4v1H0V4h8zM0 8h8V7H0v1zm0 3h8v-1H0v1z"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <a
        className="header-container"
        href="/about"
      >
        <div
          className="header-switch-theme"
        >
          <div
            className="toggle-switch"
          >
            <input
              className="toggle-switch-checkbox"
              defaultChecked={true}
              id="toggle-switch-theme"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="toggle-switch-label"
              htmlFor="toggle-switch-theme"
            >
              <span
                className="toggle-switch-inner"
                data-no="üåô"
                data-yes="‚òÄÔ∏è"
              />
            </label>
          </div>
        </div>
        üè† about
      </a>
    </div>
    <div
      className="page-container"
    >
      <h3
        className="todo-title"
      >
        Sat Nov 14 2020
      </h3>
      <button
        className="todo button-hover-light todo-false"
        id={0}
        onClick={[Function]}
      >
        <span
          className="todo-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-white)",
                "height": "3rem",
                "padding": ".5rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 416 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M272 96c26.51 0 48-21.49 48-48S298.51 0 272 0s-48 21.49-48 48 21.49 48 48 48zM113.69 317.47l-14.8 34.52H32c-17.67 0-32 14.33-32 32s14.33 32 32 32h77.45c19.25 0 36.58-11.44 44.11-29.09l8.79-20.52-10.67-6.3c-17.32-10.23-30.06-25.37-37.99-42.61zM384 223.99h-44.03l-26.06-53.25c-12.5-25.55-35.45-44.23-61.78-50.94l-71.08-21.14c-28.3-6.8-57.77-.55-80.84 17.14l-39.67 30.41c-14.03 10.75-16.69 30.83-5.92 44.86s30.84 16.66 44.86 5.92l39.69-30.41c7.67-5.89 17.44-8 25.27-6.14l14.7 4.37-37.46 87.39c-12.62 29.48-1.31 64.01 26.3 80.31l84.98 50.17-27.47 87.73c-5.28 16.86 4.11 34.81 20.97 40.09 3.19 1 6.41 1.48 9.58 1.48 13.61 0 26.23-8.77 30.52-22.45l31.64-101.06c5.91-20.77-2.89-43.08-21.64-54.39l-61.24-36.14 31.31-78.28 20.27 41.43c8 16.34 24.92 26.89 43.11 26.89H384c17.67 0 32-14.33 32-32s-14.33-31.99-32-31.99z"
            />
          </svg>
        </span>
        <span
          className="todo-item"
        >
          EXERCISE
        </span>
      </button>
      <button
        className="todo button-hover-light todo-false"
        id={1}
        onClick={[Function]}
      >
        <span
          className="todo-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-white)",
                "height": "3rem",
                "padding": ".5rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256.47 216.77l86.73 109.18s-16.6 102.36-76.57 150.12C206.66 523.85 0 510.19 0 510.19s3.8-23.14 11-55.43l94.62-112.17c3.97-4.7-.87-11.62-6.65-9.5l-60.4 22.09c14.44-41.66 32.72-80.04 54.6-97.47 59.97-47.76 163.3-40.94 163.3-40.94zM636.53 31.03l-19.86-25c-5.49-6.9-15.52-8.05-22.41-2.56l-232.48 177.8-34.14-42.97c-5.09-6.41-15.14-5.21-18.59 2.21l-25.33 54.55 86.73 109.18 58.8-12.45c8-1.69 11.42-11.2 6.34-17.6l-34.09-42.92 232.48-177.8c6.89-5.48 8.04-15.53 2.55-22.44z"
            />
          </svg>
        </span>
        <span
          className="todo-item"
        >
          CLEAN APT
        </span>
      </button>
      <button
        className="todo button-hover-light todo-false"
        id={2}
        onClick={[Function]}
      >
        <span
          className="todo-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-white)",
                "height": "3rem",
                "padding": ".5rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M639.4 433.6c-8.4-20.4-31.8-30.1-52.2-21.6l-22.1 9.2-38.7-101.9c47.9-35 64.8-100.3 34.5-152.8L474.3 16c-8-13.9-25.1-19.7-40-13.6L320 49.8 205.7 2.4c-14.9-6.2-32-.3-40 13.6L79.1 166.5C48.9 219 65.7 284.3 113.6 319.2L74.9 421.1l-22.1-9.2c-20.4-8.5-43.7 1.2-52.2 21.6-1.7 4.1.2 8.8 4.3 10.5l162.3 67.4c4.1 1.7 8.7-.2 10.4-4.3 8.4-20.4-1.2-43.8-21.6-52.3l-22.1-9.2L173.3 342c4.4.5 8.8 1.3 13.1 1.3 51.7 0 99.4-33.1 113.4-85.3l20.2-75.4 20.2 75.4c14 52.2 61.7 85.3 113.4 85.3 4.3 0 8.7-.8 13.1-1.3L506 445.6l-22.1 9.2c-20.4 8.5-30.1 31.9-21.6 52.3 1.7 4.1 6.4 6 10.4 4.3L635.1 444c4-1.7 6-6.3 4.3-10.4zM275.9 162.1l-112.1-46.5 36.5-63.4 94.5 39.2-18.9 70.7zm88.2 0l-18.9-70.7 94.5-39.2 36.5 63.4-112.1 46.5z"
            />
          </svg>
        </span>
        <span
          className="todo-item"
        >
          WASH DISHES
        </span>
      </button>
      <button
        className="todo button-hover-light todo-false"
        id={3}
        onClick={[Function]}
      >
        <span
          className="todo-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-white)",
                "height": "3rem",
                "padding": ".5rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M192 256c61.9 0 112-50.1 112-112S253.9 32 192 32 80 82.1 80 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C51.6 288 0 339.6 0 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zM480 256c53 0 96-43 96-96s-43-96-96-96-96 43-96 96 43 96 96 96zm48 32h-3.8c-13.9 4.8-28.6 8-44.2 8s-30.3-3.2-44.2-8H432c-20.4 0-39.2 5.9-55.7 15.4 24.4 26.3 39.7 61.2 39.7 99.8v38.4c0 2.2-.5 4.3-.6 6.4H592c26.5 0 48-21.5 48-48 0-61.9-50.1-112-112-112z"
            />
          </svg>
        </span>
        <span
          className="todo-item"
        >
          REACH OUT
        </span>
      </button>
      <button
        className="todo button-hover-light todo-false"
        id={4}
        onClick={[Function]}
      >
        <span
          className="todo-icon"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-white)",
                "height": "3rem",
                "padding": ".5rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M443.98 96.25c-11.01-45.22-47.11-82.06-92.01-93.72-32.19-8.36-63 5.1-89.14 24.33-3.25 2.39-6.96 3.73-10.5 5.48l28.32 18.21c7.42 4.77 9.58 14.67 4.8 22.11-4.46 6.95-14.27 9.86-22.11 4.8L162.83 12.84c-20.7-10.85-43.38-16.4-66.81-10.31-44.9 11.67-81 48.5-92.01 93.72-10.13 41.62-.42 80.81 21.5 110.43 23.36 31.57 32.68 68.66 36.29 107.35 4.4 47.16 10.33 94.16 20.94 140.32l7.8 33.95c3.19 13.87 15.49 23.7 29.67 23.7 13.97 0 26.15-9.55 29.54-23.16l34.47-138.42c4.56-18.32 20.96-31.16 39.76-31.16s35.2 12.85 39.76 31.16l34.47 138.42c3.39 13.61 15.57 23.16 29.54 23.16 14.18 0 26.48-9.83 29.67-23.7l7.8-33.95c10.61-46.15 16.53-93.16 20.94-140.32 3.61-38.7 12.93-75.78 36.29-107.35 21.95-29.61 31.66-68.8 21.53-110.43z"
            />
          </svg>
        </span>
        <span
          className="todo-item"
        >
          NIGHT HYGIENE
        </span>
      </button>
      <button
        className="todo todo-reset button-hover-light todo-resetBtn"
        id="2"
        onClick={[Function]}
      >
        <span
          className="todo-item-reset"
        >
          RESET
        </span>
        <audio
          className="checkSound"
        >
          <source
            src="check.mp3"
          />
        </audio>
        <audio
          className="successSound"
        >
          <source
            src="success.mp3"
          />
        </audio>
        <audio
          className="resetSound"
        >
          <source
            src="reset.mp3"
          />
        </audio>
      </button>
    </div>
    <div>
      <div
        className="nav nav-bottom"
      >
        <div
          className="nav-icon $
          this.props.isBottom ? \\"nav-icon-bottom\\" : \\"nav-icon-top\\"
        }"
        >
          <svg
            fill="currentColor"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": "var(--c-1)",
                "height": "3rem",
                "width": "3rem",
              }
            }
            viewBox="0 0 8 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 4v1H0V4h8zM0 8h8V7H0v1zm0 3h8v-1H0v1z"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <a
        className="footer-container"
        href="/about"
      >
        <span>
          copyright Alex Vincent-Hill 2020 - üè† about
        </span>
      </a>
    </div>
  </div>
</div>
`;
